### PR TITLE
added early stop callback

### LINKED ISSUE #19: early stopping ?
Have you considered support for stopping Mango early if the objective function doesn't improve enough?

Would you be open to a contribution PR?

thanks!

### PR DESCRIPTION
**Pull request by tihom**
Added `early_stopping` callback to allow custom logic for stopping criteria. See `test_tuner.test_early_stopping_complex` for usage example that @asah requested. Will add them to readme once PR is approved.

closes #19

### CODE REVIEW

File: mango/tuner.py [hunk 1]
```diff
@@ -60,6 +67,14 @@ def strategy_is_penalty(self):
         def strategy_is_clustering(self):
             return self.parallel_strategy == 'clustering'
 
+        def early_stop(self, results):
+            if self.early_stopping is None:
+                return False
+
+            results = copy.deepcopy(results)
```
Comment (asah) [inline mango/tuner.py:74]: is this expensive?

Comment (tihom) [reply mango/tuner.py:74]: Good point. Based on some benchmarks: it takes 1e-5 secs per optimizer iteration to do `deepcopy` on the results data structure. This orders of magnitude lower than the Bayesian optimizer overhead.

Benchmark code:

```
import numpy as np
import copy
import timeit
from scipy.stats import uniform, loguniform
from sklearn.model_selection import ParameterSampler

space = {
        'f': range(10),
        'h': uniform(-10, 20),
        'e': ['cat1', 1, 'cat2'],
        'c': [1, 2, 3],
        'a': ['const'],
        'g': loguniform(0.001, 100),
        'b': [10],
        'd': uniform(0, 1),
        'i': [True, False]
    }


iters = 1000
results = dict(
    objective_values=np.random.uniform(iters),
    params_tries=list(ParameterSampler(space, n_iter=iters)),
    surrogate_values=np.random.uniform(iters),
    best_objective=np.random.uniform(),
    best_params=list(ParameterSampler(space, n_iter=1))[0],
)


def run():
    res = copy.deepcopy(results)


n = 1000
print(timeit.timeit(run, number=n) / n)
```

Comment (asah) [reply mango/tuner.py:73]: nice!

File: mango/tuner.py [hunk 2]
```diff
@@ -36,6 +41,7 @@ class Config:
         exploration_decay: float = 0.9
         exploration_min: float = 0.1
         fixed_domain: bool = False
+        early_stopping: Callable = None
```
Comment (asah) [inline mango/tuner.py:44]: tiny code optimization to consider... if early_stopping is initialized with `lambda *_: return False` then I think you can remove the need for self.early_stop() etc and simply just call this function directly... 

obviously, totally optional, just maybe simplifies the code... obviously, if you're busy pls ignore!  I'm excited to simply start using this key feature!

Comment (tihom) [reply mango/tuner.py:44]: `lambda` would require call to `deepcopy` even when no callback is provided by the user.

**Comment by asah** [comment]
Ah makes sense. Thx!

**Pull request merged**